iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -154,7 +154,7 @@ private void ConnectionStatusChangeHandler(ConnectionStatus status, ConnectionSt
                             {
                                 // Not great to print out a connection string, but this is done for sample/demo purposes.
                                 _logger.LogWarning($"The current connection string {badCs} is invalid. Trying another.");
-                                InitializeClient();
+                                await InitializeAndOpenClientAsync();
```

Comment:
Why do we need InitializeAndOpenClientAsync at all here? It seems a little circular to set the ConnectionStatusChangeHandler within InitializeAndOpenClientAsync and call InitializeAndOpenClientAsync inside ConnectionStatusChangeHandler.

Reply:
I see your point, but in the absence of a "IsClientConnected" property on the DeviceClient, we'll have to manage the device client's lifecycle from within the status change callback. So the flow becomes:
-> initialize client -> set connection status callback handler -> use connection status callback handler to detect disconnection -> dispose the client -> initialize client (back to the start) -> ...

Do you think there might be a better way to handle this? Our current client indicates the status of operation through exceptions, and the status of the client through event handler. This does make managing the lifecycle confusing, but this is the approach I found that worked. If we can come up with something that'll be easier to implement, I would like that.

Reply:
I think I get why we are doing this. However, we are not really calling dispose before calling open right? Or am I missing something? Is it a common patter to initialize the client on error within the ConnectionStatusChangeHandler ?

Reply:
Ok I found the dispose :) Took a while to go through the flow.

Reply:
Ideally, it would be good for the DeviceClient itself to have a property to know if we need to reinitialize it. It might be better to just store s_connectionStatus and use that during Init. We don't need to call the Init here. Also if we are expecting multiple threads to hit this, the end result should be like an AND or OR of all the status so that something is not overriden.

Reply:
> We don't need to call the Init here.

Where would we handle the device client's lifetime then? The only monitoring tool that that device client gives you is this status change callback.

> the end result should be like an AND or OR of all the status so that something is not overridden.

I didn't understand what you meant here, could you give an example?

Reply:
Talked offline. We do think there is a better solution overall for this scenario but we'll do it when we re-design the device client.